articles/150_roslaunch.md

```diff
@@ -0,0 +1,364 @@
+
+- `stdout` pipe
+- `stderr` pipe
+
+The launch system may choose to either capture these pipes, for logging or suppressing output to the console, or it can connect the pipes to an existing `pty`, like the terminal's `stdout` and/or `stderr` pipes or a null pipe (e.g. `/dev/null`).
+
+When capturing the output pipes of a process, the launch system could report this data in a way that the user may process them in real-time or could pass the data through user defined filters, generating a user-handled event when the filter matches.
+
+In addition, the launch system may interact with, or allow the user to interact with, an operating system process's:
+
+- `stdin` pipe
+- signals (`SIGINT`, `SIGTERM`, `SIGUSR1`, etc...)
+
+Regardless of how the user uses the launch system to interact with these items, they should be exposed by the launch system, which is the only entity which can interact with them directly.
+
+#### Termination
+
+If the operating system process terminates, and therefore returns a return code, the launch system will report this event and it can be handled in a user defined way.
+Termination covers expected termination (e.g. return from `main()` or use `exit()`) and unexpected termination (e.g. the abort trap or a segmentation fault or bus error).
+
+Historically, ROS 1's `roslaunch` allowed a few common exit handling cases:
+
+- `require=true`: if this process exits (any reason) shutdown everything else, as it's "required"
+- `respawn=true`: if this process exits (any reason) restart it with the same settings as startup
+  - `respawn_delay=N`: if restarting it, delay a number of seconds between attempts
+
+The launch system may initiate the termination of an operating system process.
+This starts with the signaling of `SIGINT` on the child process.
+If this does not result in the termination of the process, then one of a few things can happen based on the configuration of the launch system:
+
+- after a period of time, signal `SIGTERM`
+- after a period of time, signal `SIGKILL`
+- nothing
+
+By default, the launch system will:
+
+- send `SIGINT`
+- after 10 seconds, send `SIGTERM`
+- after 10 additional seconds, send `SIGKILL`
+
+The latter two steps can be skipped, or the time until escalation can be adjusted, on a per process basis.
+
+The launch system will initiate this process when an event (built-in or user generated) initiates shutdown, e.g. when a process with the equivalent of the `require=true` exit handler terminates, or when the launch system itself receives the `SIGINT` signal.
+
+If the launch system itself receives the `SIGTERM` signal it will send the `SIGKILL` signal to all child processes and exit immediately.
+
+<div class="alert alert-warning" markdown="1">
+RFC:
+
+There are several small decisions here that I made somewhat arbitrarily, e.g. the default time until escalation and the propagation of `SIGKILL` when the launch system receives `SIGTERM`.
```

Comment:
Sending SIGKILL on receipt of SIGTERM feels a bit abrupt. Clients have no opportunity to catch it a clean up anything. Are you assuming that SIGINT was already used, and the user got impatient?

Doing the SIGTERM; wait 10 sec.; SIGKILL sequence might work better, but would not satisfy impatient users.

Reply:
Per node parameterisation of those parameters would be nice. We were at times 1) impatient, but also at times 2) requiring more shutdown time to be made available to some processes

Reply:
> Sending SIGKILL on receipt of SIGTERM feels a bit abrupt.

So to be clear, if you SIGINT the launch system, then processes run by the launch system will get SIGINT too, before being escalated to SIGTERM and then finally SIGKILL.

If you SIGTERM the launch system, it will immediately SIGKILL child processes and try to exit as quickly as possible. So only do that if you're impatient or already tried SIGINT.

If you SIGKILL the launch system, I don't know what happens to the child processes, it might be OS dependent, but it could result in "zombie" processes.

So, given that graduation of signals the user can send the launch system, I felt that these were appropriate.

> Clients have no opportunity to catch it a clean up anything.

Well, that's the idea with SIGTERM to the launch system. If you want the child processes to be able to react, then SIGINT the launch system and let it do the normal SIGINT->SIGTERM->SIGKILL escallation.

That all being said, I'm open to suggestions (and rationale) about what the launch system should do instead when receiving SIGTERM.

> Are you assuming that SIGINT was already used, and the user got impatient?

Yes, or that something is wrong in the launch system itself, so I was aiming for an "as simple as possible" SIGTERM signal handler for the launch system, which basically sends SIGKILL to all child processes and then exits itself immediately.

Doing a SIGTERM to child processes, waiting 10 seconds or until all child processes exit, and then sending SIGKILL to all child process still up might be more complicated and therefore more likely to also hang.

Basically, you'd want to remove any incentive to SIGKILL the launch system by making SIGTERM reliable and quick to exit.

> Doing the SIGTERM; wait 10 sec.; SIGKILL sequence might work better, but would not satisfy impatient users.

That might be better, I honestly don't know. Anyone else want to weigh in on that?

----

> Per node parameterisation of those parameters would be nice. We were at times 1) impatient, but also at times 2) requiring more shutdown time to be made available to some processes

Yeah, that's a good idea, and I was thinking of that myself, though I'll have to double check the document to see if that's called out clearly.

However, one concern is that it might break some expected behavior of the launch system for users. For example, if a user expects the launch system to exit and escalate at set time intervals, but some nodes aren't affected by this and instead have different schedules, it might be very confusing to them why the launch system is suddenly behaving differently or taking so long to shutdown. So we'll have to be careful about logging what's happening when extra shutdown constraints are made on a per process basis.

Reply:
Aye, +1 to clear logging on shutdown. I'd be unconfused if roslaunch printed a table every few seconds on shutdown reporting out on what processes remain and when the next escalation is due to occur.

Reply:
> > Are you assuming that SIGINT was already used, and the user got impatient?

> Yes, or that something is wrong in the launch system itself, so I was aiming for an "as simple as possible" SIGTERM signal handler for the launch system, which basically sends SIGKILL to all child processes and then exits itself immediately.
>
> Basically, you'd want to remove any incentive to SIGKILL the launch system by making SIGTERM reliable and quick to exit.

That seems reasonable to me, and the rationale you give would help other readers.

Reply:
I tried to wrap this up in 2964a67.